job/replay_validator.go

```diff
@@ -0,0 +1,133 @@
+}
+
+func NewReplayValidator(scheduler models.SchedulerUnit) *Validator {
+	return &Validator{scheduler: scheduler}
+}
+
+func (v *Validator) Validate(ctx context.Context, replaySpecRepo store.ReplaySpecRepository,
+	reqInput models.ReplayRequest, replayTree *tree.TreeNode) error {
+	if !reqInput.Force {
+		reqReplayNodes := replayTree.GetAllNodes()
+
+		//check if this dag have running instance in the scheduler
+		if err := v.validateRunningInstance(ctx, reqReplayNodes, reqInput); err != nil {
+			return err
+		}
+
+		//check another replay active for this dag
+		activeReplaySpecs, err := replaySpecRepo.GetByStatus(ReplayStatusToValidate)
+		if err != nil {
+			if err == store.ErrResourceNotFound {
+				return nil
+			}
+			return err
+		}
+		return validateReplayJobsConflict(activeReplaySpecs, reqInput, reqReplayNodes)
+	}
+	//check and cancel if found conflicted replays for same job ID
+	return cancelConflictedReplays(replaySpecRepo, reqInput)
+}
+
+func cancelConflictedReplays(replaySpecRepo store.ReplaySpecRepository, reqInput models.ReplayRequest) error {
+	duplicatedReplaySpecs, err := replaySpecRepo.GetByJobIDAndStatus(reqInput.Job.ID, ReplayStatusToValidate)
+	if err != nil {
+		if err == store.ErrResourceNotFound {
+			return nil
+		}
+		return err
+	}
+	for _, replaySpec := range duplicatedReplaySpecs {
+		if err := replaySpecRepo.UpdateStatus(replaySpec.ID, models.ReplayStatusCancelled, models.ReplayMessage{
+			Type:    ErrConflictedJobRun.Error(),
+			Message: fmt.Sprintf("force started replay with ID: %s", reqInput.ID),
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (v *Validator) validateRunningInstance(ctx context.Context, reqReplayNodes []*tree.TreeNode, reqInput models.ReplayRequest) error {
+	for _, reqReplayNode := range reqReplayNodes {
+		batchEndDate := reqInput.End.AddDate(0, 0, 1).Add(time.Second * -1)
+		jobStatusAllRuns, err := v.scheduler.GetDagRunStatus(ctx, reqInput.Project, reqReplayNode.Data.(models.JobSpec).Name, reqInput.Start, batchEndDate, schedulerBatchSize)
+		if err != nil {
+			return err
+		}
+		for _, jobStatus := range jobStatusAllRuns {
+			if reqReplayNode.Runs.Contains(jobStatus.ScheduledAt) && jobStatus.State == models.JobStatusStateRunning {
+				return ErrConflictedJobRun
+			}
+		}
+	}
+	return nil
+}
+
+func validateReplayJobsConflict(activeReplaySpecs []models.ReplaySpec, reqInput models.ReplayRequest,
+	reqReplayNodes []*tree.TreeNode) error {
+	for _, activeSpec := range activeReplaySpecs {
+		activeReplayWorkerRequest := models.ReplayRequest{
+			ID:         activeSpec.ID,
+			Job:        activeSpec.Job,
+			Start:      activeSpec.StartDate,
+			End:        activeSpec.EndDate,
+			Project:    reqInput.Project,
+			JobSpecMap: reqInput.JobSpecMap,
+		}
+		activeTree, err := prepareReplayExecutionTree(activeReplayWorkerRequest)
+		if err != nil {
+			return err
+		}
+		activeNodes := activeTree.GetAllNodes()
+		return checkAnyConflictedDags(activeNodes, reqReplayNodes)
```

Comment:
This for loop will run only once instead of iterating over all `activeReplaySpecs`. Is this expected?

Reply:
have updated this + added a test case to check this